### PR TITLE
reference: handle combination of tag and digest in ImageDelete

### PR DESCRIPTION
**Pull request by vdemeester**
If you remove an image with digest+tag, it will fail because it wont
find it in the reference store (where digest+tag -> digest). Let's
make sure we do the same in ImageDelete, stripping the tag if
digest+tag are present.

This is related to https://github.com/docker/docker/pull/30889 and should hopefully fix the `win2lin` build failures 👼

/cc @aaronlehmann @johnstep @dmcgowan 

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

**Comment by vdemeester** [comment]
https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/31617/console

**Comment by vdemeester** [comment]
Failure is weird but doesn't seems related 

```
13:49:19 ----------------------------------------------------------------------
13:49:19 FAIL: docker_cli_build_test.go:360: DockerSuite.TestBuildCacheAdd
13:49:19 
13:49:19 docker_cli_build_test.go:370:
13:49:19     buildImageSuccessfully(c, name, withDockerfile(fmt.Sprintf(`FROM scratch
13:49:19     		ADD %s/robots.txt /`, server.URL())))
13:49:19 C:/Jenkins/gopath/src/github.com/docker/docker/pkg/testutil/cmd/command.go:64:
13:49:19     t.Fatalf("at %s:%d\n%s", filepath.Base(file), line, err.Error())
13:49:19 ... Error: at docker_utils_test.go:629
13:49:19 
13:49:19 Command: C:\CI\CI-5998452\docker.exe build -t testbuildtwoimageswithadd -
13:49:19 ExitCode: 1, Error: exit status 1
13:49:19 Stdout: Sending build context to Docker daemon  2.048kB

13:49:19 Step 1/2 : FROM scratch
13:49:19  ---> 
13:49:19 Step 2/2 : ADD http://10.0.0.12:32777/robots.txt /
13:49:19 
13:49:19 Stderr: Get http://10.0.0.12:32777/robots.txt: dial tcp 10.0.0.12:32777: getsockopt: connection refused
13:49:19 
13:49:19 
13:49:19 Failures:
13:49:19 ExitCode was 1 expected 0
13:49:19 Expected no error
13:49:19 
13:49:19 
13:49:20 
13:49:20 ----------------------------------------------------------------------
```

**Review by johnstep — APPROVED** [review]
LGTM

**Comment by aaronlehmann** [comment]
I think this should actually go in the reference store's `Delete` method.

**Comment by dmcgowan** [comment]
I agree with @aaronlehmann, we are trying to keep as much of the application logic from needing to worry about whether the reference contains tag/digest. Right now the limitation seems to be in support by the reference store so we should keep logic there when possible.

**Comment by vdemeester** [comment]
hum ok, I'll see what I can do :angel:

**Comment by vdemeester** [comment]
@aaronlehmann @dmcgowan updated 👼

File: reference/store.go [hunk 1]
```diff
@@ -106,20 +106,29 @@ func (store *store) AddDigest(ref reference.Canonical, id digest.Digest, force b
 	return store.addReference(ref, id, force)
 }
 
-func (store *store) addReference(ref reference.Named, id digest.Digest, force bool) error {
+func dropTagIfPresent(originalRef reference.Named) (reference.Named, error) {
```
Comment (aaronlehmann) [inline reference/store.go]: nit: This only drops the tag if there is also a digest. `favorDigest`?

Comment (vdemeester) [inline reference/store.go]: Damn I'm bad with names… I'll update the name 👼

**Comment by vdemeester** [comment]
Updated 👼

**Comment by vdemeester** [comment]
And win2lin is green \o/

**Comment by aaronlehmann** [comment]
LGTM

**Comment by tonistiigi** [comment]
LGTM

**Pull request merged**